File: src/main/java/net/jgp/books/spark/ch17/lab250_feed_delta_population/FeedFrancePopDeltaLakeApp.java
package net.jgp.books.spark.ch17.lab250_feed_delta_population;

import static org.apache.spark.sql.functions.col;
import static org.apache.spark.sql.functions.regexp_replace;
import static org.apache.spark.sql.functions.when;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;

/**
 * 
 * @author jgp
 */
public class FeedFrancePopDeltaLakeApp {

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    FeedFrancePopDeltaLakeApp app =
        new FeedFrancePopDeltaLakeApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Load France's population dataset and store it in Delta")
        .master("local[*]")
        .getOrCreate();

    // Reads a CSV file, called population_dept.csv, stores it in a
    // dataframe
    Dataset<Row> df = spark.read().format("csv")
        .option("inferSchema", true)
        .option("header", true)
        .load("data/france_population_dept/population_dept.csv");

    df = df
        .withColumn("Code département",
            when(col("Code département").$eq$eq$eq("2A"), "20")
                .otherwise(col("Code département")))
        .withColumn("Code département",
            when(col("Code département").$eq$eq$eq("2B"), "20")
                .otherwise(col("Code département")))
        .withColumn("Code département",
            col("Code département").cast(DataTypes.IntegerType))
        .withColumn("Population municipale",
            regexp_replace(col("Population municipale"), ",", ""))
        .withColumn("Population municipale",
            col("Population municipale").cast(DataTypes.IntegerType))
        .withColumn("Population totale",
            regexp_replace(col("Population totale"), ",", ""))
        .withColumn("Population totale",
            col("Population totale").cast(DataTypes.IntegerType))
        .drop("_c9");
    df.show(25);
    df.printSchema();

    df.write().format("delta")
        .mode("overwrite")
        .save("/tmp/delta_france_population");

    System.out.println(df.count() + " rows updated.");
  }
}

File: src/main/python/lab250_feed_delta_population/feedFrancePopDeltaLakeApp.py
"""
  Export data into Delta Lake

  @author rambabu.posa
"""
import os
import logging
from pyspark.sql import (SparkSession, functions as F)

def get_absolute_file_path(path, filename):
    current_dir = os.path.dirname(__file__)
    relative_path = f"{path}{filename}"
    absolute_file_path = os.path.join(current_dir, relative_path)
    return absolute_file_path

def main(spark):
    path = "../../../../data/france_population_dept/"
    filename = "population_dept.csv"
    absolute_file_path = get_absolute_file_path(path, filename)

    # Reads a CSV file, called population_dept.csv,
    # stores it in a dataframe
    df = spark.read.format("csv") \
        .option("inferSchema", True) \
        .option("header", True) \
        .load(absolute_file_path)

    df = df.withColumn("Code département",
              F.when(F.col("Code département") == F.lit("2A"), "20")
              .otherwise(F.col("Code département"))) \
        .withColumn("Code département",
              F.when(F.col("Code département") == F.lit("2B"), "20")
               .otherwise(F.col("Code département"))) \
        .withColumn("Code département",
              F.col("Code département").cast("int")) \
        .withColumn("Population municipale",
              F.regexp_replace(F.col("Population municipale"), ",", "")) \
        .withColumn("Population municipale",
              F.col("Population municipale").cast("int")) \
        .withColumn("Population totale",
              F.regexp_replace(F.col("Population totale"), ",", "")) \
        .withColumn("Population totale",
              F.col("Population totale").cast("int")) \
        .drop("_c9")

    df.show(25)
    df.printSchema()

    df.write.format("delta") \
        .mode("overwrite") \
        .option("encoding", "utf-8") \
        .save("/tmp/delta_france_population")

    logging.warning("{} rows updated.".format(df.count()))

if __name__ == "__main__":
    # Creates a session on a local master
    # To use Databricks Delta Lake, we should add delta core packages to SparkSession
    spark = SparkSession.builder \
        .appName("Load France's population dataset and store it in Delta") \
        .config("spark.jars.packages", "io.delta:delta-core_2.12:0.7.0") \
        .master("local[*]").getOrCreate()

    # setting log level, update this as per your requirement
    spark.sparkContext.setLogLevel("warn")

    main(spark)
    spark.stop()

